atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientImpl.java

```diff
@@ -140,34 +139,28 @@ private PreparedStatement prepareStatement(String queryString) {
         }
     }
 
-    private class CqlQueryBuilderImpl<R> implements CqlQueryBuilder<R> {
+    private class CqlQueryBuilderImpl<R> implements CqlQueryBuilder<R, ListenableFuture<R>> {
 
-        private String queryString;
+        private CqlQuerySpec<R> cqlQuerySpec;
         private Map<String, Object> args = new HashMap<>();
-        private RowStreamAccumulator<R> rowStreamAccumulator;
 
         @Override
-        public CqlQueryBuilder<R> setQueryString(String query) {
-            this.queryString = query;
+        public CqlQueryBuilder<R, ListenableFuture<R>> setQuerySpec(CqlQuerySpec<R> querySpec) {
+            this.cqlQuerySpec = querySpec;
             return this;
         }
 
         @Override
-        public CqlQueryBuilder<R> setArg(String argumentName, Object argument) {
+        public CqlQueryBuilder<R, ListenableFuture<R>> setArg(String argumentName, Object argument) {
             args.put(argumentName, argument);
             return this;
         }
 
         @Override
-        public <T> CqlQueryBuilder<T> setResultSetVisitor(RowStreamAccumulator<T> visitor) {
-            this.rowStreamAccumulator = (RowStreamAccumulator<R>) visitor;
-            return (CqlQueryBuilder<T>) this;
-        }
-
-        @Override
-        public CqlQuery<R> build() {
-            Preconditions.checkNotNull(queryString, "Empty query string");
-            PreparedStatement statement = CqlClientImpl.this.prepareStatement(queryString);
+        public  CqlQuery<? extends ListenableFuture<R>> build() {
+            Preconditions.checkNotNull(cqlQuerySpec, "Empty query spec");
+            PreparedStatement statement = cache.cacheQuerySpec(cqlQuerySpec, () ->
```

Comment:
nit: would prefer if the cache was "transparent" in the sense that I have some sort of `StatementPreparer` and I just ask for it to `prepare` given a `querySpec`. Canonical impl would call the underlying `session.prepare`. Wrap that with a cache which takes in a `StatementPreparer`

Reply:
That's basically what I tried to do with the above mentioned `Supplier`

Reply:
Transparent cache.